opentelemetry-exporter-cloud-trace/tests/test_cloud_trace_exporter.py

```diff
@@ -182,28 +185,102 @@ def test_extract_status(self):
+                dropped_attributes_count=1,
+            ),
         )
 
+    def test_too_many_attributes(self):
         too_many_attrs = {}
         for attr_key in range(5):
             too_many_attrs[str(attr_key)] = 0
         proto_attrs = _extract_attributes(too_many_attrs, 4)
         self.assertEqual(proto_attrs.dropped_attributes_count, 1)
 
-    def test_extract_events(self):
+    def test_add_agent_attribute(self):
+        self.assertEqual(
+            _extract_attributes({}, 4, add_agent_attr=True),
+            ProtoSpan.Attributes(
+                attribute_map={"g.co/agent": self.agent_code},
+                dropped_attributes_count=0,
+            ),
+        )
+        # Drop existing attributes in favor of the agent attribute
+        self.assertEqual(
+            _extract_attributes(
+                {"attribute_key": "attr_value"}, 1, add_agent_attr=True
+            ),
+            ProtoSpan.Attributes(
+                attribute_map={"g.co/agent": self.agent_code},
+                dropped_attributes_count=1,
+            ),
+        )
+
+    def test_extract_empty_events(self):
         self.assertIsNone(_extract_events([]))
+
+    def test_too_many_events(self):
+        time_in_ns = 1589919268850900051
+        time_in_ms_and_ns = {"seconds": 1589919268, "nanos": 850899968}
+        event = Event(name="event", timestamp=time_in_ns, attributes={})
+        too_many_events = [event] * (MAX_NUM_EVENTS + 5)
+        self.assertEqual(
+            _extract_events(too_many_events),
+            ProtoSpan.TimeEvents(
+                time_event=[
+                    {
+                        "time": time_in_ms_and_ns,
+                        "annotation": {
+                            "description": TruncatableString(value="event",),
+                            "attributes": {},
+                        },
+                    },
+                ]
+                * MAX_NUM_EVENTS,
+                dropped_annotations_count=len(too_many_events)
+                - MAX_NUM_EVENTS,
+            ),
+        )
+
+    def test_too_many_event_attributes(self):
+        time_in_ns1 = 1589919268850900051
+        event_attrs = {}
+        for attr_key in range(MAX_EVENT_ATTRS + 5):
+            event_attrs[str(attr_key)] = 0
+        proto_events = _extract_events(
+            [Event(name="a", attributes=event_attrs, timestamp=time_in_ns1)]
+        )
+        self.assertEqual(
+            len(
+                proto_events.time_event[0].annotation.attributes.attribute_map
+            ),
+            MAX_EVENT_ATTRS,
+        )
+        self.assertEqual(
+            proto_events.time_event[
+                0
+            ].annotation.attributes.dropped_attributes_count,
+            len(event_attrs) - MAX_EVENT_ATTRS,
+        )
+
+    def test_extract_multiple_events(self):
         time_in_ns1 = 1589919268850900051
         time_in_ms_and_ns1 = {"seconds": 1589919268, "nanos": 850899968}
```

Comment:
Same comment here about numbers not matching.

Reply:
This is addressed in PR #42.